### DISCUSSION TITLE
Vector/Tensor operations in MOOSE

### DISCUSSION THREAD

**Discussion by RbmIjpb** [idaholab/moose#28111] [Q&A Modules: Solid mechanics] [answered]
Hello, 

I'm trying to run a model for a NeoHookean fiber reinforced material where the fibers are represented by vectors and the Structural Tensor is obtained by the dyadic product between the fiber vectors. 
To do so, I tried to implement some methods to handle the tensor/vector operations mainely multiplication, dyadic product and dot product between the vector. 
When I run my code I get "Segmentation Fault error". I added the .C file in .txt format (I hope that's okay).
[AnticlinalCw.txt](https://github.com/user-attachments/files/16181397/AnticlinalCw.txt)

 The debugging message is below. I'm also trying to multiply the vector by a tensor trace, but each time i get an error that this type of vectors cannot be multiplied by a double. 
I would like to know if there is an alternative way in MOOSE to do it, or if my code can be saved. 
Thank you for your help!

```
 `Thread 1 "lollipop-dbg" received signal SIGSEGV, Segmentation fault.
0x00007ffff38d4acb in AnticlinalCw::multiply_vector_tensor (this=0x555555c90aa0, a=..., tensor=...)
    at /home/antilope/projects/lollipop/src/kernels/AnticlinalCw.C:61
61            result[i] += tensor(i, j) * a[j];
(gdb) where
#0  0x00007ffff38d4acb in AnticlinalCw::multiply_vector_tensor (this=0x555555c90aa0, a=..., tensor=...)
    at /home/antilope/projects/lollipop/src/kernels/AnticlinalCw.C:61
#1  0x00007ffff38da6ac in AnticlinalCw::computeResidual (this=0x555555c90aa0,
    effective_trial_stress=@0x7fffffff50d8: -0, scalar=@0x7fffffff5128: 0)
    at /home/antilope/projects/lollipop/src/kernels/AnticlinalCw.C:322
#2  0x00007ffff38d8987 in AnticlinalCw::computeQpPK1Stress (this=0x555555c90aa0)
    at /home/antilope/projects/lollipop/src/kernels/AnticlinalCw.C:242
#3  0x00007ffff2ed5a99 in ComputeLagrangianStressPK1::computeQpStressUpdate (this=0x555555c90aa0)
    at /home/antilope/projects/moose/modules/solid_mechanics/src/materials/lagrangian/ComputeLagrangianStressPK1.C:33
#4  0x00007ffff2ed49b3 in ComputeLagrangianStressBase::computeQpProperties (this=0x555555c90aa0)
    at /home/antilope/projects/moose/modules/solid_mechanics/src/materials/lagrangian/ComputeLagrangianStressBase.C:44
#5  0x00007ffff68ebc7a in Material::computeProperties (this=0x555555c90aa0)
    at /home/antilope/projects/moose/framework/src/materials/Material.C:139
#6  0x00007ffff6641891 in MaterialData::reinit<std::vector<std::shared_ptr<MaterialBase>, std::allocator<std::shared_ptr<MaterialBase> > > > (this=0x555555b27300, mats=...)
    at /home/antilope/projects/moose/framework/build/header_symlinks/MaterialData.h:313
#7  0x00007ffff65d89f6 in FEProblemBase::reinitMaterials (this=0x555555b23310, blk_id=0, tid=0, swap_stateful=true)
    at /home/antilope/projects/moose/framework/src/problems/FEProblemBase.C:3528
#8  0x00007ffff5c49975 in ComputeUserObjectsThread::onElement (this=0x7fffffffaa40, elem=0x555555af1370)
    at /home/antilope/projects/moose/framework/src/loops/ComputeUserObjectsThread.C:126
#9  0x00007ffff4a0228d in ThreadedElementLoopBase<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >::operator() (this=0x7fffffffaa40, range=..., bypass_threading=false)
    at /home/antilope/projects/moose/framework/build/header_symlinks/ThreadedElementLoopBase.h:248
#10 0x00007ffff6644e1b in libMesh::Threads::parallel_reduce<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*>, ComputeUserObjectsThread> (range=..., body=...)
    at /home/antilope/miniforge/envs/moose/libmesh/include/libmesh/threads_pthread.h:370
#11 0x00007ffff65e1b67 in FEProblemBase::computeUserObjectsInternal(MooseEnumItem const&, Moose::AuxGroup const&, TheWarehouse::QueryCache<>&) (this=0x555555b23310, type=..., group=@0x7fffffffafd0: Moose::POST_AUX, primary_query=...)
    at /home/antilope/projects/moose/framework/src/problems/FEProblemBase.C:4375`
```

**Comment by GiudGiud** [comment]
Hello

since it s a seg fault you are likely accessing out of bounds, either using a debugger or a print statement, make sure
result, tensor and a are of the correct sizes

**Comment by RbmIjpb** [comment]
Thank you for you answer. It seems that the tensor and result are correctly read, but if i do vector a.size() i get 0 , and i can't access the coordinates of a. I don't understand why it's happening. 


here's the code and and an example : 
`std::vector<double> AnticlinalCw::multiply_vector_tensor(const std::vector<double> & a, const RankTwoTensor & tensor)
{
  constexpr unsigned int N = Moose::dim; // Use the dimension of the tensor
  std::vector<double> result(N, 0.0); // Initialize the result vector with zeros

  for (unsigned int i = 0; i < N; ++i)
  {
    for (unsigned int j = 0; j < N; ++j)
    {
      result[i] += tensor(i, j) * a[j];
    }
  }

  return result;

}
`


`std::vector<double> sa = multiply_vector_tensor(_a, s); `



`0x00007ffff38d4acf in AnticlinalCw::multiply_vector_tensor (this=0x555555c90aa0, a=..., tensor=...)
    at /home/antilope/projects/lollipop/src/kernels/AnticlinalCw.C:61
61            result[i] += tensor(i, j) * a[j];
(gdb) print a
$1 = (const
    std::vector<double, std::allocator<double> > &) @0x555555c92410: {<std::_Vector_base<double, std::allocator<double> >> = {
    _M_impl = {<std::allocator<double>> = {<__gnu_cxx::new_allocator<double>> = {<No data fields>}, <No data fields>}, <std::_Vector_base<double, std::allocator<double> >::_Vector_impl_data> = {_M_start = 0x0, _M_finish = 0x0,
        _M_end_of_storage = 0x0}, <No data fields>}}, <No data fields>}
(gdb) print result
$2 = {<std::_Vector_base<double, std::allocator<double> >> = {
    _M_impl = {<std::allocator<double>> = {<__gnu_cxx::new_allocator<double>> = {<No data fields>}, <No data fields>}, <std::_Vector_base<double, std::allocator<double> >::_Vector_impl_data> = {_M_start = 0x555555dd84e0,
        _M_finish = 0x555555dd84f8, _M_end_of_storage = 0x555555dd84f8}, <No data fields>}}, <No data fields>}
(gdb) print tensor
$3 = (const RankTwoTensor &) @0x555555c924a0: {<libMesh::TensorValue<double>> = {<libMesh::TypeTensor<double>> = {
      _coords = {0, 0, 0, 0, 0, 0, 0, 0, 0}}, <No data fields>}, static N = 3, static N2 = 9, static identityCoords = {
    1, 0, 0, 0, 1, 0, 0, 0, 1}}
(gdb) print a.size()
$4 = 0
(gdb) print tensor ._coords
$5 = {0, 0, 0, 0, 0, 0, 0, 0, 0}
(gdb) print result.size()
Currently Executing
  Performing Initial Setup
    Computing User Objects$6 = 3`

**Reply by GiudGiud** [reply]
what is a ? how is it retrieved?
you can look where `multiply_vector_tensor(` is being called since a is an argument of that function

**Comment by RbmIjpb** [comment]
a is an argument of the multiply_vector_tensor( ; I tried to verify whether the size of a was correct in the function. 
_a is std::vector<double> in the code and s is a rank two tensor.
in this code i tried to implement _a = _a0/(A.trace()) but it didn't work so i'm just running it as _a = _a0
here's how _a0 is retrieved : _a0 (getParam<std::vector<double>>("fiber_vector"))

**Reply by GiudGiud** [reply]
what is "fiber_vector" in the input? Please paste the whole object
what are _a and _a0  in the header ? (.h file)

**Comment by RbmIjpb** [comment]
Here's the .C file
[AnticlinalCw.txt](https://github.com/user-attachments/files/16182343/AnticlinalCw.txt)
Here's the .h file
[AnticlinalCw.txt](https://github.com/user-attachments/files/16182350/AnticlinalCw.txt)
and Here is the .i file 
[AnticlinalCw.txt](https://github.com/user-attachments/files/16182365/AnticlinalCw.txt)

**Reply by GiudGiud** [reply]
_a is defined both in the source file
```
  std::vector<Real> _a = _a0 ; // 1/A.trace() * _a0 didn't work 
```
and in the header as a `std::vector<double> _a` attribute.

please make sure this does not happen

**Comment by RbmIjpb** [comment]
Thank you for the clarification. I corrected the code and still have the Segmentation Fault problem : 
h. file
[AnticlinalCw.txt](https://github.com/user-attachments/files/16183587/AnticlinalCw.txt)

C. file
[AnticlinalCw.txt](https://github.com/user-attachments/files/16183588/AnticlinalCw.txt)

**Reply by GiudGiud** [reply, marked as answer]
it is not fixed I still see 
```
  std::vector<double> _a = _a0 ; // 1/A.trace() * _a0 didn't work 
```
in the code (the .C file, in computeQpPK1Stress())

_a, the attribute, the one that causes the seg fault, is never set

**Reply by RbmIjpb** [reply]
Problem Solved ! Thank you very much !
